src/develop/imageop.c

```diff
@@ -2083,6 +2090,11 @@ void dt_iop_commit_params(dt_iop_module_t *module,
 
     free(str);
   }
+
+  if(module->operation_tags() & IOP_TAG_DISTORT)
+  {
+    DT_DEBUG_CONTROL_SIGNAL_RAISE(darktable.signals, DT_SIGNAL_DEVELOP_DISTORT);
```

Comment:
It feels to me completely _wrong_ to raise a signal in `commit_params` (which is basically the last step of preparing the data to calculate the pipe) that needs to be handled in the _gui_ and there might decide that the whole current pipe run is useless anyway and needs to be redone. It is another sign that there is something fundamentally insufficient in the current pipe & cache setup that causes aggressive reruns and cache purges. Rather than putting duck tape everywhere this should really be tackled structurally.. All these "fixes" everywhere makes that increasingly difficult and unlikely.

Reply:
> It feels to me completely _wrong_ to raise a signal in `commit_params`

That's not the final version, just a "working" version I cooked in the morning. This will be fixed.

> It is another sign that there is something fundamentally insufficient in the current pipe & cache setup that causes aggressive reruns and cache purges.

No that's the opposite. The work from @jenshannoschwalm has fixed the pipe caching and this now do far less computation but it has broken the tone-equalizer which was working by chance as taking a preview snapshot for the luminance mask when `process()` was run. And now that `process()` is called only when we miss data in the cache tone-equalizer is using an old/obsolete mask.

Reply:
The point with iop cache is, the  module's parameters must reflect the full status of processing. If we have an internal mask that is changed and which is not reflected in params (there are used for the hash) we used to full reprocess. I see two ways:

1) make sure we have something in params reflecting this
2) if we change some internal mask and can't guarantee a correct hash, we can invalidate the cache lines later than the module and thus avoid calculating previous modules (later and  previous are meant as positions in the pipeline)

Reply:
Btw, that's how the internal histogram  for some modules were fixed.

Reply:
@jenshannoschwalm : I think there is miss understanding. The pipe is ok. The processing is ok.

But the tone-equalizer does keep an internal copy of the "cache" (preview buffer). So when you change a module doing distortion you end up with a wrong internal cache not representing what you have displayed in darkroom.

The solution proposed here can be summarized as:
- When a module doing distortion is enabled/disabled raise a signal (DT_SIGNAL_DEVELOP_DISTORT).
- Tone equalizer is connected to this signal and when it is activated and as focus it trigger a preview re-computation to be able to get a new cache as luminance mask.

That's basically what this PR does, nothing more. I still need to review it but the summary above should be accurate.

This is very specific to tone-equalizer which is the only one having on-canvas picker to display the luminance and select the proper cursor to activate.

I see the following possibilities:
1. Accept this PR which is doing nothing more when tone-equalizer is not used.
2. Remove the on-canvas picker and let user interact with tone-equalizer with the sliders and/or curve.  
3. Keep it broken as it is in current master
4. A better solution that I don't see at this stage :)

Reply:
> But the tone-equalizer does keep an internal copy of the "cache" (preview buffer). So when you change a module doing distortion you end up with a wrong internal cache not representing what you have displayed in darkroom

So it takes the whole pixelpipe output of preview buffer? That's surprising for me but would probably  require all data to be processed again. I thought that
1. The distorter might be earlier in the pipe. In that case it would have either params reflected in it's hash (good case) or not; if so it would require the distorter to invalidate all cachelines after this module. see `dt_dev_pixelpipe_cache_invalidate_later()`
2. If the distorter is later in the pipe everything should be good.

I looked into the code and now i understand your intention and the way it works. Seems good to me .